In [1]:
# 1. Встановлюємо інші потрібні бібліотеки (без pgserver)
%pip install -q psycopg2-binary sqlalchemy pandas pyarrow ucimlrepo

# 2. Встановлюємо та запускаємо класичний PostgreSQL сервер у системі Colab
!sudo apt-get -y -qq update
!sudo apt-get -y -qq install postgresql
!sudo service postgresql start

# 3. Налаштовуємо пароль для стандартного користувача postgres
!sudo -u postgres psql -c "ALTER USER postgres PASSWORD 'postgres';"

# 4. Підключаємося через SQLAlchemy
import pandas as pd
from sqlalchemy import create_engine, text

# Використовуємо стандартний рядок підключення до нашого локального сервера
db_uri = 'postgresql+psycopg2://postgres:postgres@localhost:5432/postgres'
engine = create_engine(db_uri, future=True)

# Перевіряємо підключення
with engine.connect() as conn:
    version = conn.execute(text('SELECT version()')).scalar_one()
    print("Сервер успішно запущено!")
    print(version)

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 39.4 MB/s eta 0:00:00
W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu noble InRelease' does not seem to provide it (sources.list entry misspelt?)
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 79, <STDIN> line 12.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Selecting previously unselected package libjson-perl.
(Reading database ... 122809 files and directories currently installed.)
Preparing to unpack .../00-libjson-perl_4.10000-1_all.deb ...
Unpacking libjson-perl (4.10000-1) ...
Selecting previously unselected package postgresql-

In [2]:
from ucimlrepo import fetch_ucirepo
import pandas as pd

# Автоматично тягнемо датасет Adult (id=2) прямо з сервера UCI
dataset = fetch_ucirepo(id=2)
df = pd.concat([dataset.data.features, dataset.data.targets], axis=1)

# Зберігаємо його у локальний CSV всередині машини Colab
csv_path = '/content/hw3_dataset.csv'
df.to_csv(csv_path, index=False)

print(f"Розмір датасету: {df.shape}")
print(f"Перші рядки збережено!")

Розмір датасету: (48842, 15)
Перші рядки збережено!


In [3]:
from sqlalchemy import text
import pandas as pd

# 1. Створюємо staging-таблицю з усіма колонками у форматі TEXT
create_staging_query = text("""
DROP TABLE IF EXISTS hw3_adult_clean CASCADE;
DROP TABLE IF EXISTS hw3_adult_staging CASCADE;

CREATE TABLE hw3_adult_staging (
    age_raw TEXT,
    workclass_raw TEXT,
    fnlwgt_raw TEXT,
    education_raw TEXT,
    education_num_raw TEXT,
    marital_status_raw TEXT,
    occupation_raw TEXT,
    relationship_raw TEXT,
    race_raw TEXT,
    sex_raw TEXT,
    capital_gain_raw TEXT,
    capital_loss_raw TEXT,
    hours_per_week_raw TEXT,
    native_country_raw TEXT,
    income_raw TEXT
);
""")

with engine.begin() as conn:
    conn.execute(create_staging_query)

# 2. Заливаємо дані через COPY FROM STDIN
raw_conn = engine.raw_connection()
try:
    with raw_conn.cursor() as cur:
        # Відкриваємо наш щойно створений CSV
        with open('/content/hw3_dataset.csv', 'r', encoding='utf-8', newline='') as file:
            cur.copy_expert(
                """
                COPY hw3_adult_staging (
                    age_raw, workclass_raw, fnlwgt_raw, education_raw, education_num_raw,
                    marital_status_raw, occupation_raw, relationship_raw, race_raw, sex_raw,
                    capital_gain_raw, capital_loss_raw, hours_per_week_raw, native_country_raw, income_raw
                )
                FROM STDIN
                WITH (
                    FORMAT csv,
                    HEADER true,
                    DELIMITER ','
                )
                """,
                file
            )
    raw_conn.commit()
finally:
    raw_conn.close()

# 3. Перевіряємо кількість завантажених рядків
loaded_rows = pd.read_sql('SELECT COUNT(*) AS loaded_rows FROM hw3_adult_staging', engine)
print("Дані успішно завантажено в staging-таблицю!")
print(loaded_rows)

Дані успішно завантажено в staging-таблицю!
   loaded_rows
0        48842


In [4]:
from sqlalchemy import text

transform_query = text("""
-- 1. Створюємо чисту типізовану таблицю
DROP TABLE IF EXISTS hw3_adult_clean CASCADE;

CREATE TABLE hw3_adult_clean (
    adult_id         BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,

    age              SMALLINT NOT NULL
                     CHECK (age > 0 AND age <= 120), -- Предметне обмеження 1

    workclass        TEXT,
    fnlwgt           INTEGER,
    education        TEXT,

    education_num    SMALLINT
                     CHECK (education_num > 0 AND education_num <= 25), -- Предметне обмеження 2

    marital_status   TEXT,
    occupation       TEXT,
    relationship     TEXT,
    race             TEXT,
    sex              TEXT,
    capital_gain     INTEGER,
    capital_loss     INTEGER,

    hours_per_week   SMALLINT
                     CHECK (hours_per_week > 0 AND hours_per_week <= 168), -- Предметне обмеження 3

    native_country   TEXT,
    income           TEXT
);

-- 2. Очищаємо та переносимо дані
WITH parsed AS (
    SELECT
        NULLIF(TRIM(age_raw), '?')::SMALLINT AS age,
        NULLIF(TRIM(workclass_raw), '?') AS workclass,
        NULLIF(TRIM(fnlwgt_raw), '?')::INTEGER AS fnlwgt,
        NULLIF(TRIM(education_raw), '?') AS education,
        NULLIF(TRIM(education_num_raw), '?')::SMALLINT AS education_num,
        NULLIF(TRIM(marital_status_raw), '?') AS marital_status,
        NULLIF(TRIM(occupation_raw), '?') AS occupation,
        NULLIF(TRIM(relationship_raw), '?') AS relationship,
        NULLIF(TRIM(race_raw), '?') AS race,
        NULLIF(TRIM(sex_raw), '?') AS sex,
        NULLIF(TRIM(capital_gain_raw), '?')::INTEGER AS capital_gain,
        NULLIF(TRIM(capital_loss_raw), '?')::INTEGER AS capital_loss,
        NULLIF(TRIM(hours_per_week_raw), '?')::SMALLINT AS hours_per_week,
        NULLIF(TRIM(native_country_raw), '?') AS native_country,
        NULLIF(TRIM(income_raw), '?') AS income
    FROM hw3_adult_staging
)
INSERT INTO hw3_adult_clean (
    age, workclass, fnlwgt, education, education_num,
    marital_status, occupation, relationship, race, sex,
    capital_gain, capital_loss, hours_per_week, native_country, income
)
SELECT
    age, workclass, fnlwgt, education, education_num,
    marital_status, occupation, relationship, race, sex,
    capital_gain, capital_loss, hours_per_week, native_country, income
FROM parsed
WHERE age IS NOT NULL; -- Рядок без віку не має сенсу для цього датасету
""")

with engine.begin() as conn:
    conn.execute(transform_query)

# Перевіряємо, чи успішно перелилися дані
clean_rows = pd.read_sql('SELECT COUNT(*) AS clean_rows FROM hw3_adult_clean', engine)
print("Дані успішно очищено та перенесено!")
print(clean_rows)

Дані успішно очищено та перенесено!
   clean_rows
0       48842


# Обґрунтування типів даних та правил очищення (Adult Dataset)

1.   Ідентифікатор (adult_id)

          *   Семантика: Унікальний ключ запису.
          *   Джерельний формат: Відсутній у сирих даних.
          *   Вибраний тип: BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY.
          *   Очищення: Генерується базою автоматично під час вставки.

2.   Числові показники (age, fnlwgt, education_num, capital_gain, capital_loss, hours_per_week)

          *   Семантика: Вік, вага запису, тривалість навчання, фінансові показники та робочі години.
          *   Джерельний формат: Сирий TEXT.
          *   Вибраний тип: INTEGER (для великих чисел на кшталт ваги чи капіталу) та SMALLINT (для віку й годин, щоб економити пам'ять).
          *   Очищення: NULLIF(TRIM(col), '?')::INTEGER. Функція TRIM прибирає зайві пробіли, а NULLIF перетворює специфічний для цього датасету маркер відсутності даних ? на справжній SQL NULL.

3.   Категоріальні та текстові ознаки (workclass, education, marital_status, occupation, relationship, race, sex, native_country, income)

          *   Семантика: Робочий клас, освіта, сімейний стан, професія, стать, країна та цільова змінна доходу.
          *   Джерельний формат: Сирий TEXT.
          *   Вибраний тип: TEXT. У PostgreSQL немає сенсу використовувати VARCHAR(255), оскільки TEXT є еталонним типом для рядків без жорстких бізнес-обмежень.
          *   Очищення: NULLIF(TRIM(col), '?'). Прибираємо пробіли по краях і конвертуємо маркери ? у NULL.









In [5]:
audit_1_query = """
SELECT 'staging' AS layer, COUNT(*) AS n_rows FROM hw3_adult_staging
UNION ALL
SELECT 'clean' AS layer, COUNT(*) AS n_rows FROM hw3_adult_clean
ORDER BY layer DESC;
"""
print(pd.read_sql(audit_1_query, engine))

     layer  n_rows
0  staging   48842
1    clean   48842


**3.1. Контроль кількості рядків:**

Кількість рядків у staging та clean таблицях збігається (48842). Фільтр WHERE age IS NOT NULL під час перенесення не відкинув жодного рядка, оскільки сирий датасет не містив пропусків у колонці віку.

In [6]:
audit_2_query = """
SELECT
    COUNT(*) AS total_rows,
    COUNT(DISTINCT (
        age, workclass, education, marital_status,
        occupation, race, sex, native_country
    )) AS unique_business_keys,
    COUNT(*) - COUNT(DISTINCT (
        age, workclass, education, marital_status,
        occupation, race, sex, native_country
    )) AS potential_duplicates
FROM hw3_adult_clean;
"""
print(pd.read_sql(audit_2_query, engine))

   total_rows  unique_business_keys  potential_duplicates
0       48842                 27118                 21724


**3.2. Перевірка дублікатів:**

Оскільки датасет не має природного ключа (наприклад, SSN), як business_key було обрано комбінацію демографічних та робочих характеристик: age, workclass, education, marital_status, occupation, race, sex, native_country. Різниця між загальною кількістю рядків та унікальними комбінаціями показує наявність потенційних дублікатів (людей з абсолютно ідентичними анкетами).

In [7]:
audit_3_query = """
SELECT
    ROUND(COUNT(*) FILTER (WHERE workclass IS NULL) * 100.0 / COUNT(*), 2) AS pct_null_workclass,
    ROUND(COUNT(*) FILTER (WHERE occupation IS NULL) * 100.0 / COUNT(*), 2) AS pct_null_occupation,
    ROUND(COUNT(*) FILTER (WHERE native_country IS NULL) * 100.0 / COUNT(*), 2) AS pct_null_country,
    ROUND(COUNT(*) FILTER (WHERE capital_gain IS NULL) * 100.0 / COUNT(*), 2) AS pct_null_cap_gain,
    ROUND(COUNT(*) FILTER (WHERE hours_per_week IS NULL) * 100.0 / COUNT(*), 2) AS pct_null_hours
FROM hw3_adult_clean;
"""
print(pd.read_sql(audit_3_query, engine))

   pct_null_workclass  pct_null_occupation  pct_null_country  \
0                5.73                 5.75              1.75   

   pct_null_cap_gain  pct_null_hours  
0                0.0             0.0  


**3.3. Null-rate аудит:**

Найбільший відсоток пропусків (NULL) спостерігається в категоріальних ознаках occupation та workclass. Числові колонки (age, hours_per_week) заповнені на 100%. Відсутність професії часто корелює з відсутністю робочого класу, що є логічним для тих, хто ніколи не працював.

In [8]:
task_4_1_query = """
SELECT
    income,
    COUNT(*) AS n_rows,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2) AS pct_rows
FROM hw3_adult_clean
GROUP BY income
ORDER BY n_rows DESC;
"""
print(pd.read_sql(task_4_1_query, engine))

   income  n_rows  pct_rows
0   <=50K   24720     50.61
1  <=50K.   12435     25.46
2    >50K    7841     16.05
3   >50K.    3846      7.87


**4.1. Розподіл за категорією income:**

У датасеті спостерігається сильний дисбаланс класів: більшість записів (близько 76%) належать до групи з доходом <=50K, і лише 24% — до >50K. Цей дисбаланс є критично важливим для машинного навчання, оскільки модель може почати ігнорувати менший клас. При навчанні це потрібно буде компенсувати (наприклад, через зважування класів або oversampling).

In [9]:
task_4_2_query = """
SELECT
    COUNT(hours_per_week) AS n_non_null,
    MIN(hours_per_week) AS min_value,
    MAX(hours_per_week) AS max_value,
    ROUND(AVG(hours_per_week)::NUMERIC, 2) AS avg_value,
    ROUND(STDDEV_SAMP(hours_per_week)::NUMERIC, 2) AS sd_value
FROM hw3_adult_clean;
"""
print(pd.read_sql(task_4_2_query, engine))

   n_non_null  min_value  max_value  avg_value  sd_value
0       48842          1         99      40.42     12.39


**4.2. Статистика колонки hours_per_week:**

Середній робочий тиждень становить близько 40 годин, що відповідає стандартному робочому графіку. Однак мінімальне значення — 1 година, а максимальне — 99 годин на тиждень (що є типовим порогом відсікання "top-coding" у переписах населення). Стандартне відхилення становить близько 12.3 годин.

In [10]:
task_4_3_query = """
WITH stats AS (
    SELECT
        AVG(age) AS mean_value,
        STDDEV_SAMP(age) AS sd_value
    FROM hw3_adult_clean
    WHERE age IS NOT NULL
)
SELECT
    t.adult_id,
    t.age,
    t.workclass,
    t.income,
    ROUND(((t.age - s.mean_value) / NULLIF(s.sd_value, 0))::NUMERIC, 2) AS z_score
FROM hw3_adult_clean AS t
CROSS JOIN stats AS s
WHERE t.age IS NOT NULL
  AND s.sd_value > 0
  AND ABS(t.age - s.mean_value) > 3 * s.sd_value
ORDER BY z_score DESC
LIMIT 20;
"""
print(pd.read_sql(task_4_3_query, engine))

    adult_id  age         workclass income  z_score
0      11515   90           Private  <=50K     3.75
1      10548   90           Private   >50K     3.75
2       8976   90           Private   >50K     3.75
3       6236   90  Self-emp-not-inc  <=50K     3.75
4       8966   90              None  <=50K     3.75
5      10213   90  Self-emp-not-inc  <=50K     3.75
6       2307   90           Private  <=50K     3.75
7       5374   90         Local-gov   >50K     3.75
8       2896   90           Private  <=50K     3.75
9        223   90           Private  <=50K     3.75
10      6628   90           Private  <=50K     3.75
11      5410   90           Private   >50K     3.75
12      1938   90           Private  <=50K     3.75
13      8810   90           Private   >50K     3.75
14      4075   90           Private  <=50K     3.75
15      4114   90              None  <=50K     3.75
16      1041   90           Private  <=50K     3.75
17      5109   90           Private  <=50K     3.75
18      5276

In [11]:
q1 = """
SELECT
    occupation,
    ROUND(AVG(capital_gain), 2) AS avg_capital_gain
FROM hw3_adult_clean
WHERE occupation IS NOT NULL
GROUP BY occupation
ORDER BY avg_capital_gain DESC
LIMIT 5;
"""
print(pd.read_sql(q1, engine))

        occupation  avg_capital_gain
0   Prof-specialty           2745.92
1  Exec-managerial           2277.76
2            Sales           1266.90
3     Craft-repair            718.95
4  Protective-serv            717.07


**Запит 1 (GROUP BY, ORDER BY, LIMIT)**

*Які 5 професій мають найвищий середній капітальний прибуток (capital gain)?*

**Інтерпретація:** Найбільший прибуток від капіталу мають професії рівня "Prof-specialty" та "Exec-managerial". Це логічно, оскільки керівні та висококваліфіковані посади зазвичай передбачають опціони на акції або інвестиційні бонуси.


In [12]:
q2 = """
SELECT COUNT(*) AS hard_working_educated_women
FROM hw3_adult_clean
WHERE sex = 'Female'
  AND (education = 'Bachelors' OR education = 'Masters')
  AND hours_per_week > 40;
"""
print(pd.read_sql(q2, engine))

   hard_working_educated_women
0                          930


**Запит 2 (WHERE з AND/OR та дужками)**

*Скільки жінок із вищою освітою (Bachelors або Masters) працюють понад 40 годин на тиждень?*

**Інтерпретація:** Запит повертає специфічний сегмент висококваліфікованих жінок, які перепрацьовують понад стандартну норму. Ця група може бути цікавою для аналізу гендерного розриву в оплаті праці на високих посадах.

In [13]:
q3 = """
SELECT
    workclass,
    ROUND(AVG(age), 1) AS avg_age
FROM hw3_adult_clean
WHERE workclass IN ('Private', 'Local-gov', 'State-gov')
GROUP BY workclass
ORDER BY avg_age DESC;
"""
print(pd.read_sql(q3, engine))

   workclass  avg_age
0  Local-gov     41.7
1  State-gov     39.5
2    Private     36.9


**Запит 3 (Оператор IN)**

*Який середній вік працівників у приватному, місцевому та державному секторах?*

**Інтерпретація:** Запит демонструє використання оператора IN для фільтрації конкретних класів. Працівники державного сектора в середньому дещо старші за працівників приватного сектора, що може свідчити про вищу стабільність та меншу плинність кадрів у держустановах.

In [16]:
q4 = """
SELECT
    occupation,
    COUNT(*) AS workers_count
FROM hw3_adult_clean
WHERE occupation ILIKE '%%repair%%'
GROUP BY occupation;
"""
print(pd.read_sql(q4, engine))

     occupation  workers_count
0  Craft-repair           6112


**Запит 4 (Оператор LIKE)**

*Скільки людей працюють у сфері ремонту та крафтового виробництва (професії, що містять слово 'repair')?*

**Інтерпретація:** Використання ILIKE дозволило знайти всі варіації професій, пов'язаних із ремонтом та ремеслом (Craft-repair). Ця група складає значну частину вибірки синіх комірців, технічні навички яких (як от пайка чи збірка складних механізмів) високо цінуються на ринку.

In [17]:
q5 = """
SELECT
    marital_status,
    COUNT(*) AS n_rows
FROM hw3_adult_clean
WHERE age BETWEEN 25 AND 45
GROUP BY marital_status
ORDER BY n_rows DESC;
"""
print(pd.read_sql(q5, engine))

          marital_status  n_rows
0     Married-civ-spouse   12716
1          Never-married    7660
2               Divorced    3912
3              Separated     992
4  Married-spouse-absent     361
5                Widowed     197
6      Married-AF-spouse      28


**Запит 5 (Оператор BETWEEN)**

*Який розподіл сімейного стану серед людей активного робочого віку (від 25 до 45 років)?*

**Інтерпретація:** У найактивнішому робочому віці найбільшу частку становлять одружені особи та ті, хто ніколи не був у шлюбі. Оператор BETWEEN чітко відсік студентську молодь та пенсіонерів, залишаючи ядро робочої сили.

In [18]:
q6 = """
SELECT
    COUNT(*) FILTER (WHERE native_country IS NULL) AS null_countries,
    ROUND(COUNT(*) FILTER (WHERE native_country IS NULL) * 100.0 / COUNT(*), 2) AS pct_null
FROM hw3_adult_clean;
"""
print(pd.read_sql(q6, engine))

   null_countries  pct_null
0             857      1.75


**Запит 6 (IS NULL / IS NOT NULL)**

*Яка частка людей без вказаної країни походження (native_country є NULL)?*

**Інтерпретація:** Відсоток людей з невідомою країною походження є вкрай малим (менше 2%). Оскільки цей відсоток незначний, для побудови ML-моделі такі пропуски можна заповнити найчастішим значенням (модою) — "United-States", або ж замінити на окрему категорію "Unknown".

In [19]:
q7 = """
SELECT
    COALESCE(workclass, 'Unknown_Sector') AS filled_workclass,
    COUNT(*) AS n_rows
FROM hw3_adult_clean
GROUP BY filled_workclass
ORDER BY n_rows DESC;
"""
print(pd.read_sql(q7, engine))

   filled_workclass  n_rows
0           Private   33906
1  Self-emp-not-inc    3862
2         Local-gov    3136
3    Unknown_Sector    2799
4         State-gov    1981
5      Self-emp-inc    1695
6       Federal-gov    1432
7       Without-pay      21
8      Never-worked      10


**Запит 7 (COALESCE)**

*Як виглядатиме розподіл за робочим класом, якщо всі пропуски (NULL) замінити на категорію "Unknown_Sector"?*

**Інтерпретація:** Функція COALESCE успішно замінила всі NULL-значення на запасний рядок. Ми бачимо, що група "Unknown_Sector" займає помітне місце у вибірці, підтверджуючи необхідність окремої обробки таких ознак перед подачею в модель машинного навчання.

In [20]:
q8 = """
SELECT
    workclass,
    SUM(capital_gain) AS total_gain,
    SUM(capital_loss) AS total_loss,
    ROUND((SUM(capital_gain)::NUMERIC / NULLIF(SUM(capital_loss), 0)), 2) AS gain_loss_ratio
FROM hw3_adult_clean
WHERE workclass IS NOT NULL
GROUP BY workclass
ORDER BY gain_loss_ratio DESC NULLS LAST;
"""
print(pd.read_sql(q8, engine))

          workclass  total_gain  total_loss  gain_loss_ratio
0      Self-emp-inc     8700086      281742            30.88
1  Self-emp-not-inc     6881098      422300            16.29
2           Private    30384366     2738536            11.10
3         State-gov     1498302      163830             9.15
4       Federal-gov     1322148      155922             8.48
5         Local-gov     2503245      320261             7.82
6       Without-pay        6830        1887             3.62
7      Never-worked           0           0              NaN


**Запит 8 (NULLIF)**

*Яке співвідношення між прибутком (capital_gain) та втратами (capital_loss) для кожного робочого класу, і як уникнути ділення на нуль?*

**Інтерпретація:** Більшість людей мають нульові фінансові втрати. Завдяки функції NULLIF(capital_loss, 0) ми перетворили нулі у знаменнику на NULL, що врятувало запит від критичної помилки "division by zero" і дозволило коректно порахувати метрику для тих, у кого втрати все ж були.

In [21]:
q9 = """
SELECT
    COUNT(*) AS not_private_or_unknown
FROM hw3_adult_clean
WHERE workclass IS DISTINCT FROM 'Private';
"""
print(pd.read_sql(q9, engine))

   not_private_or_unknown
0                   14936


**Запит 9 (IS DISTINCT FROM)**

*кільки записів НЕ належать до приватного сектора ('Private'), враховуючи навіть ті, де робочий клас взагалі не вказаний (NULL)?*

**Інтерпретація:** Оператор IS DISTINCT FROM дозволив нам коректно відфільтрувати дані. На відміну від звичайного workclass != 'Private', який би повністю викинув рядки з NULL, цей безпечний оператор залишив у вибірці людей без вказаного класу, вважаючи їх "відмінними" від приватного.

In [22]:
q10 = """
SELECT DISTINCT
    education,
    sex
FROM hw3_adult_clean
WHERE income IN ('>50K', '>50K.')
ORDER BY education, sex;
"""
print(pd.read_sql(q10, engine))

       education     sex
0           10th  Female
1           10th    Male
2           11th  Female
3           11th    Male
4           12th  Female
5           12th    Male
6        1st-4th    Male
7        5th-6th  Female
8        5th-6th    Male
9        7th-8th  Female
10       7th-8th    Male
11           9th  Female
12           9th    Male
13    Assoc-acdm  Female
14    Assoc-acdm    Male
15     Assoc-voc  Female
16     Assoc-voc    Male
17     Bachelors  Female
18     Bachelors    Male
19     Doctorate  Female
20     Doctorate    Male
21       HS-grad  Female
22       HS-grad    Male
23       Masters  Female
24       Masters    Male
25     Preschool    Male
26   Prof-school  Female
27   Prof-school    Male
28  Some-college  Female
29  Some-college    Male


**Запит 10 (DISTINCT та множинний SELECT)**

*Які унікальні комбінації освіти та статі присутні серед людей, що заробляють понад 50 тисяч (включаючи брудні мітки '>50K.')?*

**Інтерпретація:** Запит DISTINCT повертає вичерпний перелік комбінацій освіти та статі для групи з високим доходом. Це допомагає швидко зрозуміти, чи представлені обидві статі на всіх рівнях освіти серед найбільш оплачуваних працівників.


---



# Reflection

1.   Проблеми якості даних:  
Головною проблемою стала наявність специфічних маркерів ? замість стандартних SQL NULL у категоріальних ознаках (наприклад, workclass, occupation). Також через відсутність унікального ідентифікатора (як-от SSN) довелося створювати складений бізнес-ключ для виявлення потенційних дублікатів. Найцікавішим знайденим артефактом стали "брудні" категорії у цільовій змінній income (значення з крапкою >50K. та <=50K.), які з'явилися через механічне злиття оригінального тренувального та тестового наборів даних.

1.   Відбір ознак (ML-features):  
Для моделювання варто обов'язково використати age, education_num, occupation, hours_per_week, оскільки вони мають прямий вплив на рівень заробітку. Натомість колонку fnlwgt (final weight) потрібно виключити — це статистична вага вибірки для демографічних розрахунків, а не характеристика конкретної людини, тому алгоритм може знайти хибні кореляції. Текстову колонку education також можна відкинути, оскільки education_num є її числовим порядковим еквівалентом.


2.   Ризик витоку даних (Leakage):  
Яскраво вираженого leakage тут немає, проте capital_gain (приріст капіталу) та capital_loss можуть розраховуватися вже за підсумками звітного податкового року. Якщо бізнес-мета полягає в прогнозуванні доходу на початку року на основі лише анкетних даних, використання цих колонок може створити частковий витік інформації з майбутнього.



2.   Рекомендації для Data Engineer:  
Для покращення ETL-пайплайну необхідно додати крок автоматичної нормалізації маркерів ? у NULL ще до потрапляння в базу. Також варто написати парсер для очищення цільової колонки від зайвих крапок і генерувати сурогатний унікальний хеш-ідентифікатор для кожного рядка, щоб легше відстежувати повні дублікати.

1.   Тип ML-задачі:  
Датасету ідеально відповідає задача бінарної класифікації (Classification) — модель має навчитися відносити людину до одного з двох класів: дохід менше або дорівнює 50 тисячам доларів чи більше цієї суми.



